File: backend/src/main/java/com/horain/analytics/AnalyticsService.java
package com.horain.analytics;

import com.horain.dto.ProjectDto;
import com.horain.dto.TimeLogDto;
import com.horain.service.ProjectService;
import com.horain.service.TimeLogService;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Analytics service for time log queries.
 * Provides total duration by project, by period, recent activity, and summaries.
 */
@Service
public class AnalyticsService {

    private final TimeLogService timeLogService;
    private final ProjectService projectService;

    public AnalyticsService(TimeLogService timeLogService, ProjectService projectService) {
        this.timeLogService = timeLogService;
        this.projectService = projectService;
    }

    public int sumTimeForPeriod(Instant start, Instant end) {
        return timeLogService.sumDurationForPeriod(start, end);
    }

    public int sumTimeByProject(UUID projectId, Instant start, Instant end) {
        return timeLogService.sumDurationByProject(projectId, start, end);
    }

    public List<TimeLogDto> getRecentLogs(int limit) {
        return timeLogService.findRecentLogs(limit);
    }

    public List<TimeLogDto> getTimeLogsForPeriod(Instant start, Instant end, UUID projectId) {
        return timeLogService.findLogsForPeriod(start, end, projectId);
    }

    /**
     * Aggregates time logs for chart display.
     *
     * @param groupBy "day_and_project" for stacked bar (days on x-axis, projects as series),
     *                "project_only" for pie (distribution by project)
     * @return Map with "categories" (List of strings) and "series" (List of {name, data})
     *         Values in data are hours (decimal).
     */
    public Map<String, Object> getTimeAggregatedForChart(Instant start, Instant end, String groupBy, ZoneId zone) {
        List<TimeLogDto> logs = timeLogService.findLogsForPeriod(start, end, null);
        Map<UUID, String> projectNames = projectService.findAll().stream()
                .collect(Collectors.toMap(ProjectDto::getId, ProjectDto::getName));

        if ("project_only".equals(groupBy)) {
            Map<UUID, Integer> projectToMinutes = new LinkedHashMap<>();
            for (TimeLogDto log : logs) {
                projectToMinutes.merge(log.getProjectId(), log.getDurationMinutes(), Integer::sum);
            }
            List<String> categories = new ArrayList<>();
            List<Double> data = new ArrayList<>();
            for (Map.Entry<UUID, Integer> e : projectToMinutes.entrySet()) {
                categories.add(projectNames.getOrDefault(e.getKey(), "?"));
                data.add(Math.round(e.getValue() / 6.0) / 10.0);
            }
            return Map.of(
                    "categories", categories,
                    "series", List.of(Map.of("name", "Heures", "data", data)));
        }

        if ("day_and_project".equals(groupBy)) {
            Map<LocalDate, Map<UUID, Integer>> dayToProjectToMinutes = new TreeMap<>();
            for (TimeLogDto log : logs) {
                LocalDate day = log.getLoggedAt().atZone(zone).toLocalDate();
                dayToProjectToMinutes
                        .computeIfAbsent(day, k -> new HashMap<>())
                        .merge(log.getProjectId(), log.getDurationMinutes(), Integer::sum);
            }
            List<LocalDate> days = new ArrayList<>(dayToProjectToMinutes.keySet());
            Set<UUID> allProjects = logs.stream().map(TimeLogDto::getProjectId).collect(Collectors.toSet());
            List<UUID> projectsOrdered = allProjects.stream()
                    .sorted((a, b) -> {
                        int totalA = logs.stream().filter(l -> l.getProjectId().equals(a)).mapToInt(TimeLogDto::getDurationMinutes).sum();
                        int totalB = logs.stream().filter(l -> l.getProjectId().equals(b)).mapToInt(TimeLogDto::getDurationMinutes).sum();
                        return Integer.compare(totalB, totalA);
                    })
                    .toList();

            DateTimeFormatter dayFormat = DateTimeFormatter.ofPattern("EEE d", java.util.Locale.FRENCH);
            List<String> categories = days.stream()
                    .map(d -> d.format(dayFormat))
                    .toList();

            List<Map<String, Object>> series = new ArrayList<>();
            for (UUID projectId : projectsOrdered) {
                List<Double> data = new ArrayList<>();
                for (LocalDate day : days) {
                    int minutes = dayToProjectToMinutes.getOrDefault(day, Map.of()).getOrDefault(projectId, 0);
                    data.add(Math.round(minutes / 6.0) / 10.0);
                }
                series.add(Map.<String, Object>of("name", projectNames.getOrDefault(projectId, "?"), "data", data));
            }

            return Map.of("categories", categories, "series", series);
        }

        return Map.of("categories", List.of(), "series", List.of());
    }

    /**
     * Start of today in the given timezone (inclusive).
     */
    public static Instant startOfDay(ZoneId zone) {
        return LocalDate.now(zone).atStartOfDay(zone).toInstant();
    }

    /**
     * End of today in the given timezone (exclusive, i.e. start of next day).
     */
    public static Instant endOfDay(ZoneId zone) {
        return LocalDate.now(zone).plusDays(1).atStartOfDay(zone).toInstant();
    }

    /**
     * Start of the week (Monday) in the given timezone.
     */
    public static Instant startOfWeek(ZoneId zone) {
        LocalDate now = LocalDate.now(zone);
        LocalDate monday = now.with(java.time.DayOfWeek.MONDAY);
        return monday.atStartOfDay(zone).toInstant();
    }

    /**
     * End of the week (Sunday 23:59:59.999) - exclusive bound for next week start.
     */
    public static Instant endOfWeek(ZoneId zone) {
        return startOfWeek(zone).plus(java.time.Duration.ofDays(7));
    }

    /**
     * Start of the month in the given timezone.
     */
    public static Instant startOfMonth(ZoneId zone) {
        LocalDate now = LocalDate.now(zone);
        LocalDate first = now.withDayOfMonth(1);
        return first.atStartOfDay(zone).toInstant();
    }

    /**
     * End of the month (exclusive).
     */
    public static Instant endOfMonth(ZoneId zone) {
        LocalDate now = LocalDate.now(zone);
        LocalDate first = now.withDayOfMonth(1);
        return first.plusMonths(1).atStartOfDay(zone).toInstant();
    }
}

File: backend/src/main/kotlin/com/horain/analytics/AnalyticsService.kt
package com.horain.analytics

import com.horain.dto.TimeLogDto
import com.horain.service.ProjectService
import com.horain.service.TimeLogService
import org.springframework.stereotype.Service
import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale
import java.util.UUID
import kotlin.math.roundToInt

/**
 * Analytics service for time log queries.
 * Provides total duration by project, by period, recent activity, and summaries.
 */
@Service
class AnalyticsService(
    private val timeLogService: TimeLogService,
    private val projectService: ProjectService
) {

    fun sumTimeForPeriod(start: Instant, end: Instant): Int =
        timeLogService.sumDurationForPeriod(start, end)

    fun sumTimeByProject(projectId: UUID, start: Instant, end: Instant): Int =
        timeLogService.sumDurationByProject(projectId, start, end)

    fun sumBillableTimeForPeriod(start: Instant, end: Instant): Int =
        timeLogService.sumDurationForPeriodByBillable(start, end, true)

    fun sumNonBillableTimeForPeriod(start: Instant, end: Instant): Int =
        timeLogService.sumDurationForPeriodByBillable(start, end, false)

    fun getRecentLogs(limit: Int): List<TimeLogDto> =
        timeLogService.findRecentLogs(limit)

    fun getTimeLogsForPeriod(start: Instant, end: Instant, projectId: UUID?): List<TimeLogDto> =
        timeLogService.findLogsForPeriod(start, end, projectId)

    /**
     * Aggregates time logs for chart display.
     *
     * @param groupBy "day_and_project" for stacked bar (days on x-axis, projects as series),
     *                "project_only" for pie (distribution by project),
     *                "billable_vs_non_billable" for pie or bar (Facturé vs Non facturé)
     * @return Map with "categories" (List of strings) and "series" (List of {name, data})
     *         Values in data are hours (decimal).
     */
    fun getTimeAggregatedForChart(
        start: Instant,
        end: Instant,
        groupBy: String,
        zone: ZoneId
    ): Map<String, Any?> {
        val logs = timeLogService.findLogsForPeriod(start, end, null)
        val projectNames = projectService.findAll().associate { it.id!! to it.name!! }

        if (groupBy == "billable_vs_non_billable") {
            val billableMinutes = timeLogService.sumDurationForPeriodByBillable(start, end, true)
            val nonBillableMinutes = timeLogService.sumDurationForPeriodByBillable(start, end, false)
            val categories = listOf("Facturé", "Non facturé")
            val data = listOf(
                roundHours(billableMinutes),
                roundHours(nonBillableMinutes)
            )
            return mapOf(
                "categories" to categories,
                "series" to listOf(mapOf("name" to "Heures", "data" to data))
            )
        }

        if (groupBy == "day_and_billable") {
            val dayToBillableNonBillable = sortedMapOf<LocalDate, IntArray>()
            for (log in logs) {
                val day = log.loggedAt!!.atZone(zone).toLocalDate()
                val pair = dayToBillableNonBillable.getOrPut(day) { intArrayOf(0, 0) }
                if (log.billable == true) {
                    pair[0] += log.durationMinutes!!
                } else {
                    pair[1] += log.durationMinutes!!
                }
            }
            val days = dayToBillableNonBillable.keys.toList()
            val dayFormat = DateTimeFormatter.ofPattern("EEE d", Locale.FRENCH)
            val categories = days.map { d -> d.format(dayFormat) }
            val billableHours = mutableListOf<Double>()
            val nonBillableHours = mutableListOf<Double>()
            for (d in days) {
                val pair = dayToBillableNonBillable[d]!!
                billableHours.add(roundHours(pair[0]))
                nonBillableHours.add(roundHours(pair[1]))
            }
            val series: List<Map<String, Any?>> = listOf(
                mapOf("name" to "Facturables", "data" to billableHours),
                mapOf("name" to "Non Facturables", "data" to nonBillableHours)
            )
            return mapOf("categories" to categories, "series" to series)
        }

        if (groupBy == "project_only") {
            val projectToMinutes = linkedMapOf<UUID, Int>()
            for (log in logs) {
                projectToMinutes.merge(log.projectId!!, log.durationMinutes!!, Int::plus)
            }
            val categories = mutableListOf<String>()
            val data = mutableListOf<Double>()
            for ((pid, minutes) in projectToMinutes) {
                categories.add(projectNames[pid] ?: "?")
                data.add(roundHours(minutes))
            }
            return mapOf(
                "categories" to categories,
                "series" to listOf(mapOf("name" to "Heures", "data" to data))
            )
        }

        if (groupBy == "day_and_project") {
            val dayToProjectToMinutes = sortedMapOf<LocalDate, MutableMap<UUID, Int>>()
            for (log in logs) {
                val day = log.loggedAt!!.atZone(zone).toLocalDate()
                val byProject = dayToProjectToMinutes.getOrPut(day) { mutableMapOf() }
                byProject.merge(log.projectId!!, log.durationMinutes!!, Int::plus)
            }
            val days = dayToProjectToMinutes.keys.toList()
            val allProjects = logs.map { it.projectId!! }.toSet()
            val projectsOrdered = allProjects.sortedWith { a, b ->
                val totalA = logs.filter { it.projectId == a }.sumOf { it.durationMinutes!! }
                val totalB = logs.filter { it.projectId == b }.sumOf { it.durationMinutes!! }
                totalB.compareTo(totalA)
            }
            val dayFormat = DateTimeFormatter.ofPattern("EEE d", Locale.FRENCH)
            val categories = days.map { d -> d.format(dayFormat) }
            val series = mutableListOf<Map<String, Any?>>()
            for (projectId in projectsOrdered) {
                val data = mutableListOf<Double>()
                for (day in days) {
                    val minutes = dayToProjectToMinutes[day]?.get(projectId) ?: 0
                    data.add(roundHours(minutes))
                }
                series.add(mapOf("name" to (projectNames[projectId] ?: "?"), "data" to data))
            }
            return mapOf("categories" to categories, "series" to series)
        }

        return mapOf("categories" to emptyList<String>(), "series" to emptyList<Any>())
    }

    companion object {
        private fun roundHours(minutes: Int): Double =
            (minutes / 6.0).roundToInt() / 10.0

        /**
         * Start of today in the given timezone (inclusive).
         */
        @JvmStatic
        fun startOfDay(zone: ZoneId): Instant =
            startOfDay(zone, LocalDate.now(zone))

        /**
         * Start of the given calendar day in the timezone (inclusive).
         */
        @JvmStatic
        fun startOfDay(zone: ZoneId, day: LocalDate): Instant =
            day.atStartOfDay(zone).toInstant()

        /**
         * End of today in the given timezone (exclusive, i.e. start of next day).
         */
        @JvmStatic
        fun endOfDay(zone: ZoneId): Instant =
            endOfDay(zone, LocalDate.now(zone))

        /**
         * End of the given calendar day in the timezone (exclusive).
         */
        @JvmStatic
        fun endOfDay(zone: ZoneId, day: LocalDate): Instant =
            day.plusDays(1).atStartOfDay(zone).toInstant()

        /**
         * Start of the week (Monday) in the given timezone.
         */
        @JvmStatic
        fun startOfWeek(zone: ZoneId): Instant =
            startOfWeek(zone, LocalDate.now(zone))

        /**
         * Start of the ISO week (Monday) containing [day] in the given timezone.
         */
        @JvmStatic
        fun startOfWeek(zone: ZoneId, day: LocalDate): Instant {
            val monday = day.with(DayOfWeek.MONDAY)
            return monday.atStartOfDay(zone).toInstant()
        }

        /**
         * End of the week (Sunday 23:59:59.999) - exclusive bound for next week start.
         */
        @JvmStatic
        fun endOfWeek(zone: ZoneId): Instant =
            endOfWeek(zone, LocalDate.now(zone))

        /**
         * End of the ISO week (exclusive) containing [day].
         */
        @JvmStatic
        fun endOfWeek(zone: ZoneId, day: LocalDate): Instant =
            startOfWeek(zone, day).plus(java.time.Duration.ofDays(7))

        /**
         * Start of the month in the given timezone.
         */
        @JvmStatic
        fun startOfMonth(zone: ZoneId): Instant =
            startOfMonth(zone, LocalDate.now(zone))

        /**
         * Start of the month containing [day].
         */
        @JvmStatic
        fun startOfMonth(zone: ZoneId, day: LocalDate): Instant {
            val first = day.withDayOfMonth(1)
            return first.atStartOfDay(zone).toInstant()
        }

        /**
         * End of the month (exclusive).
         */
        @JvmStatic
        fun endOfMonth(zone: ZoneId): Instant =
            endOfMonth(zone, LocalDate.now(zone))

        /**
         * End of the month containing [day] (exclusive).
         */
        @JvmStatic
        fun endOfMonth(zone: ZoneId, day: LocalDate): Instant {
            val first = day.withDayOfMonth(1)
            return first.plusMonths(1).atStartOfDay(zone).toInstant()
        }
    }
}
